src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1382 @@
+                  Error,
+                  kj::str("Module evaluation failed to resolve: ", specifier.getHref()));
+      // If the promise rejected, throw the rejection value.
+      if (state == v8::Promise::kRejected) {
+        js.throwException(js.v8Ref(promise->Result()));
+      }
+
+      KJ_ASSERT(state == v8::Promise::kFulfilled);
+      return module->GetModuleNamespace().As<v8::Object>();
+    };
+
+    return js.tryCatch([&]() -> v8::MaybeLocal<v8::Object> {
+      // Do we already have a cached module for this context?
+      KJ_IF_SOME(found, lookupCache.find<kj::HashIndex<ContextCallbacks>>(context)) {
+        return evaluate(js, found, context.specifier, getObserver());
+      }
+
+      KJ_IF_SOME(found, resolveWithCaching(js, context)) {
+        return evaluate(js, found, context.specifier, getObserver());
+      }
+
+      JSG_FAIL_REQUIRE(Error, kj::str("Module not found: ", context.specifier.getHref()));
+    }, [&](Value exception) {
+      // Use the isolate to rethrow the exception here instead of using the lock
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return v8::MaybeLocal<v8::Object>();
+    });
+  }
+
+  // Lookup a module that may have already been previously resolved and cached.
+  kj::Maybe<Entry&> lookup(Lock& js, v8::Local<v8::Module> module) {
+    return lookupCache.find<kj::HashIndex<EntryCallbacks>>(
+        HashableV8Ref<v8::Module>(js.v8Isolate, module))
+            .map([](Entry& entry) -> Entry& {
+      return entry;
+    });
+  }
+
+private:
+  ModuleRegistry& inner;
+  const CompilationObserver& observer;
+
+  const CompilationObserver& getObserver() const {
+    return observer;
+  }
+
+  struct EntryCallbacks {
+    const HashableV8Ref<v8::Module>& keyForRow(const Entry& entry) const {
+      return entry.key;
+    }
+    bool matches(const Entry& entry, const HashableV8Ref<v8::Module>& key) const {
+      return entry.key == key;
+    }
+    uint hashCode(const HashableV8Ref<v8::Module>& ref) const {
+      return ref.hashCode();
+    }
+  };
+
+  struct ContextCallbacks {
+    const SpecifierContext& keyForRow(const Entry& entry) const {
+      return entry.specifier;
+    }
+    bool matches(const Entry& entry, const SpecifierContext& specifier) const {
+      return entry.specifier == specifier;
+    }
+    uint hashCode(const SpecifierContext& specifier) const {
+      return specifier.hashCode();
+    }
+  };
+
+  struct UrlCallbacks {
+    const Url& keyForRow(const Entry& entry) const {
+      return entry.specifier.specifier;
+    }
+    bool matches(const Entry& entry, const Url& specifier) const {
+      return entry.specifier.specifier == specifier;
+    }
+    uint hashCode(const Url& specifier) const {
+      return specifier.hashCode();
+    }
+  };
+
+  // Resolves the module from the inner ModuleRegistry, caching the results.
+  kj::Maybe<Entry&> resolveWithCaching(Lock& js, const ResolveContext& context)
```

Comment:
Name seems a bit misleading, I suggest resolveAndCache maybe?

Reply:
Hmm, I'm not entirely sure how that's different ;-)